In [12]:
import pandas as pd
import os
import re

In [13]:
df = pd.read_csv('flipkart_smartphones_combined.csv')

In [14]:
df

,product_name,brand,current_price,original_mrp,discount_pct,rating,ratings_count,review_count,scraped_at,page_number
0,"vivo T5x 5G (Star Silver, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1
1,"vivo T5x 5G (Cyber Green, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1
2,"realme P4 Lite 5G (Mosaic Green, 64 GB)",Realme,15499,28999.0,46.0,4.3,8068,414,2026-06-30 16:02:20,1
3,"vivo T5x 5G (Cyber Green, 256 GB)",Vivo,26999,33999.0,20.0,4.4,14728,981,2026-06-30 16:02:20,1
4,"vivo T5x 5G (Cyber Green, 128 GB)",Vivo,24999,30999.0,19.0,4.4,14728,981,2026-06-30 16:02:20,1
...,...,...,...,...,...,...,...,...,...,...
1336,"MOTOROLA Edge 70 Fusion (Pantone SILHOUETTE, 2...",Motorola,31999,45999.0,30.0,4.4,12406,1181,2026-07-06 16:54:24,8
1337,"OnePlus Nord CE6 (Pitch Black, 128 GB)",OnePlus,33412,40999.0,18.0,4.5,302,17,2026-07-06 16:54:24,8
1338,"Samsung S25 Ultra 5G (Titanium Gray, 256 GB)",Samsung,89825,129999.0,30.0,4.6,558,43,2026-07-06 16:54:24,8
1339,"MOTOROLA Edge 70 Fusion (Pantone ORIENT BLUE, ...",Motorola,33999,49999.0,32.0,4.3,2450,226,2026-07-06 16:54:24,8


In [15]:
df.describe()

,current_price,original_mrp,discount_pct,rating,ratings_count,review_count,page_number
count,1341.000000,1138.000000,1128.000000,1341.000000,1341.000000,1341.000000,1341.000000
mean,25182.278896,35093.347979,25.773936,4.325652,21816.097688,1186.866518,4.496644
std,17281.197451,20888.716021,13.873601,0.161886,44642.992656,2161.132429,2.292303
min,6499.000000,8999.000000,1.000000,3.700000,14.000000,0.000000,1.000000
25%,15499.000000,22999.000000,16.000000,4.300000,1091.000000,75.000000,2.000000
50%,20999.000000,30499.000000,23.000000,4.300000,3537.000000,226.000000,4.000000
75%,29499.000000,39999.000000,40.000000,4.400000,14970.000000,1147.000000,6.000000
max,142900.000000,154900.000000,57.000000,4.800000,246757.000000,12041.000000,8.000000


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1341 entries, 0 to 1340
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_name   1341 non-null   object 
 1   brand          1341 non-null   object 
 2   current_price  1341 non-null   int64  
 3   original_mrp   1138 non-null   float64
 4   discount_pct   1128 non-null   float64
 5   rating         1341 non-null   float64
 6   ratings_count  1341 non-null   int64  
 7   review_count   1341 non-null   int64  
 8   scraped_at     1341 non-null   object 
 9   page_number    1341 non-null   int64  
dtypes: float64(3), int64(4), object(3)
memory usage: 104.9+ KB


In [17]:
df.isnull()

,product_name,brand,current_price,original_mrp,discount_pct,rating,ratings_count,review_count,scraped_at,page_number
0,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...
1336,False,False,False,False,False,False,False,False,False,False
1337,False,False,False,False,False,False,False,False,False,False
1338,False,False,False,False,False,False,False,False,False,False
1339,False,False,False,False,False,False,False,False,False,False


In [18]:
df.isnull().sum()

product_name       0
brand              0
current_price      0
original_mrp     203
discount_pct     213
rating             0
ratings_count      0
review_count       0
scraped_at         0
page_number        0
dtype: int64

In [19]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1336    False
1337    False
1338    False
1339    False
1340    False
Length: 1341, dtype: bool

In [20]:
df[df.duplicated()]

,product_name,brand,current_price,original_mrp,discount_pct,rating,ratings_count,review_count,scraped_at,page_number
6,"vivo T5x 5G (Star Silver, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1
10,"vivo T5x 5G (Cyber Green, 256 GB)",Vivo,26999,33999.0,20.0,4.4,14728,981,2026-06-30 16:02:20,1
12,"vivo T5x 5G (Cyber Green, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1
407,"realme P4R 5G (Silver Glare, 128 GB)",Realme,18999,34999.0,45.0,4.4,555,53,2026-07-02 16:07:22,1
710,"realme P4R 5G (Silver Glare, 128 GB)",Realme,18999,34999.0,45.0,4.4,580,53,2026-07-03 16:10:26,6
760,"realme P4 Lite 5G (Mosaic Blue, 64 GB)",Realme,14749,28999.0,49.0,4.3,8393,428,2026-07-03 16:10:35,8
1185,"realme P4R 5G (Silver Glare, 128 GB)",Realme,18999,34999.0,45.0,4.4,660,55,2026-07-06 16:54:04,2
1271,"realme P4R 5G (Silver Glare, 128 GB)",Realme,18999,34999.0,45.0,4.4,660,55,2026-07-06 16:54:18,6
1329,"MOTOROLA Edge 70 Fusion (Pantone SILHOUETTE, 5...",Motorola,36999,64999.0,43.0,4.3,2450,226,2026-07-06 16:54:24,8


In [21]:
df[df.duplicated(subset=["product_name", "scraped_at"])]

,product_name,brand,current_price,original_mrp,discount_pct,rating,ratings_count,review_count,scraped_at,page_number
4,"vivo T5x 5G (Cyber Green, 128 GB)",Vivo,24999,30999.0,19.0,4.4,14728,981,2026-06-30 16:02:20,1
6,"vivo T5x 5G (Star Silver, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1
10,"vivo T5x 5G (Cyber Green, 256 GB)",Vivo,26999,33999.0,20.0,4.4,14728,981,2026-06-30 16:02:20,1
12,"vivo T5x 5G (Cyber Green, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1
39,"OnePlus Nord CE6 Lite (Hyper Black, 128 GB)",OnePlus,25134,33999.0,26.0,4.3,225,14,2026-06-30 16:02:24,2
...,...,...,...,...,...,...,...,...,...,...
1283,"OPPO K13 Turbo Pro 5G (Purple Phantom, 256 GB)",OPPO,29999,41999.0,28.0,4.4,566,47,2026-07-06 16:54:18,6
1287,"OPPO K14x 5G (Prism Violet, 128 GB)",OPPO,17999,NaN,NaN,4.3,5123,271,2026-07-06 16:54:18,6
1309,"MOTOROLA Edge 70 Fusion (Pantone BLUE SURF, 25...",Motorola,31999,45999.0,30.0,4.4,12406,1181,2026-07-06 16:54:21,7
1329,"MOTOROLA Edge 70 Fusion (Pantone SILHOUETTE, 5...",Motorola,36999,64999.0,43.0,4.3,2450,226,2026-07-06 16:54:24,8


In [22]:
df["brand"].value_counts()

brand
Motorola    244
Realme      229
Samsung     173
POCO        155
Vivo        143
OPPO        108
OnePlus      79
Apple        47
Ai+          46
Redmi        46
iQOO         25
Infinix      14
LAVA         11
HMD          11
Google       10
Name: count, dtype: int64

In [23]:
df["current_price"].describe()

count      1341.000000
mean      25182.278896
std       17281.197451
min        6499.000000
25%       15499.000000
50%       20999.000000
75%       29499.000000
max      142900.000000
Name: current_price, dtype: float64

In [24]:
df["scraped_date"] = df["scraped_at"].str[:10]
conflicts = df.groupby(["product_name","scraped_date"])["current_price"].nunique()
print(conflicts[conflicts > 1].count(), "products with price conflicts")

114 products with price conflicts


In [25]:
df.groupby("scraped_date")["product_name"].nunique()

scraped_date
2026-06-30    128
2026-07-01    142
2026-07-02    141
2026-07-03    134
2026-07-04    154
2026-07-05    147
2026-07-06    112
Name: product_name, dtype: int64

### Observations
  #### original_mrp and discount_pct are in float
  #### scraped_at should be in datetime format
  #### Missing values in original_mrp,discount_pct
  #### Same product, same timestamp, different page (rows 4, 12)
  #### Rows 1271 and 1185 have same product, same date (2026-07-06) but different page numbers (2 & 6)
  #### Row 4 & 12 — same product but different prices (24999 vs 22999)
  #### Product names contain colour variants (Blue, Green, Black, etc.), which can create duplicate entries for the same model

#### converting scraped_at to datetime

In [26]:
df["scraped_at"] = pd.to_datetime(df["scraped_at"])

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1341 entries, 0 to 1340
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   product_name   1341 non-null   object        
 1   brand          1341 non-null   object        
 2   current_price  1341 non-null   int64         
 3   original_mrp   1138 non-null   float64       
 4   discount_pct   1128 non-null   float64       
 5   rating         1341 non-null   float64       
 6   ratings_count  1341 non-null   int64         
 7   review_count   1341 non-null   int64         
 8   scraped_at     1341 non-null   datetime64[ns]
 9   page_number    1341 non-null   int64         
 10  scraped_date   1341 non-null   object        
dtypes: datetime64[ns](1), float64(3), int64(4), object(3)
memory usage: 115.4+ KB


#### extracting date column needed for all day level analysis

In [28]:
df["scraped_date"] = df["scraped_at"].dt.date

In [29]:
df

,product_name,brand,current_price,original_mrp,discount_pct,rating,ratings_count,review_count,scraped_at,page_number,scraped_date
0,"vivo T5x 5G (Star Silver, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1,2026-06-30
1,"vivo T5x 5G (Cyber Green, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1,2026-06-30
2,"realme P4 Lite 5G (Mosaic Green, 64 GB)",Realme,15499,28999.0,46.0,4.3,8068,414,2026-06-30 16:02:20,1,2026-06-30
3,"vivo T5x 5G (Cyber Green, 256 GB)",Vivo,26999,33999.0,20.0,4.4,14728,981,2026-06-30 16:02:20,1,2026-06-30
4,"vivo T5x 5G (Cyber Green, 128 GB)",Vivo,24999,30999.0,19.0,4.4,14728,981,2026-06-30 16:02:20,1,2026-06-30
...,...,...,...,...,...,...,...,...,...,...,...
1336,"MOTOROLA Edge 70 Fusion (Pantone SILHOUETTE, 2...",Motorola,31999,45999.0,30.0,4.4,12406,1181,2026-07-06 16:54:24,8,2026-07-06
1337,"OnePlus Nord CE6 (Pitch Black, 128 GB)",OnePlus,33412,40999.0,18.0,4.5,302,17,2026-07-06 16:54:24,8,2026-07-06
1338,"Samsung S25 Ultra 5G (Titanium Gray, 256 GB)",Samsung,89825,129999.0,30.0,4.6,558,43,2026-07-06 16:54:24,8,2026-07-06
1339,"MOTOROLA Edge 70 Fusion (Pantone ORIENT BLUE, ...",Motorola,33999,49999.0,32.0,4.3,2450,226,2026-07-06 16:54:24,8,2026-07-06


In [30]:
df.shape

(1341, 11)

#### Product names also include storage/RAM information, which should be preserved because they represent different product variants

In [36]:
def create_base_product(product_name):
    """
    Creates a standardized product name by:
    - Keeping the phone model
    - Keeping RAM information
    - Keeping Storage information
    - Removing all colour/marketing names
    """

    if pd.isna(product_name):
        return product_name

    product_name = product_name.strip()

    # Model name = everything before first bracket
    model = product_name.split("(")[0].strip()

    # Extract RAM
    ram_match = re.search(
        r'(\d+\s*GB\s*RAM|\d+\s*GB)',
        product_name,
        re.IGNORECASE
    )

    # Extract Storage
    storage_match = re.search(
        r'(\d+\s*GB\s*(?:Storage)?)',
        product_name,
        re.IGNORECASE
    )

    # Find all GB values
    gb_values = re.findall(r'\d+\s*GB', product_name, re.IGNORECASE)

    ram = None
    storage = None

    if "RAM" in product_name.upper():

        ram = re.search(
            r'(\d+\s*GB)\s*RAM',
            product_name,
            re.IGNORECASE
        )

        ram = ram.group(1) if ram else None

        storage = re.search(
            r'(\d+\s*GB)\s*(Storage)?',
            product_name,
            re.IGNORECASE
        )

        if storage:
            storage = storage.group(1)

    else:

        if len(gb_values) >= 2:
            ram = gb_values[0]
            storage = gb_values[1]

        elif len(gb_values) == 1:
            storage = gb_values[0]

    specs = []

    if ram:
        specs.append(f"{ram} RAM")

    if storage:
        specs.append(storage)

    if specs:
        return f"{model} ({', '.join(specs)})"

    return model

In [37]:
df["base_product"] = df["product_name"].apply(create_base_product)

df[["product_name", "base_product"]].head(30)

,product_name,base_product
0,"vivo T5x 5G (Star Silver, 128 GB)",vivo T5x 5G (128 GB)
1,"vivo T5x 5G (Cyber Green, 128 GB)",vivo T5x 5G (128 GB)
2,"realme P4 Lite 5G (Mosaic Green, 64 GB)",realme P4 Lite 5G (64 GB)
3,"vivo T5x 5G (Cyber Green, 256 GB)",vivo T5x 5G (256 GB)
4,"vivo T5x 5G (Cyber Green, 128 GB)",vivo T5x 5G (128 GB)
5,"vivo T5 Pro 5G (Glacier Blue, 128 GB)",vivo T5 Pro 5G (128 GB)
6,"vivo T5x 5G (Star Silver, 128 GB)",vivo T5x 5G (128 GB)
7,"Samsung M06 5G (Sage Green, 128 GB)",Samsung M06 5G (128 GB)
8,"realme P4 5G (Steel Grey, 128 GB)",realme P4 5G (128 GB)
9,"Ai+ Pulse 2 (Purple, 64 GB)",Ai+ Pulse 2 (64 GB)


In [38]:
df_clean = df.drop_duplicates(
    subset=[
        "base_product",
        "current_price",
        "rating",
        "review_count"
    ]
)

print("Rows before :", len(df))
print("Rows after  :", len(df_clean))

Rows before : 1341
Rows after  : 654


In [40]:
df_clean

,product_name,brand,current_price,original_mrp,discount_pct,rating,ratings_count,review_count,scraped_at,page_number,scraped_date,base_product
0,"vivo T5x 5G (Star Silver, 128 GB)",Vivo,22999,28999.0,20.0,4.4,12342,703,2026-06-30 16:02:20,1,2026-06-30,vivo T5x 5G (128 GB)
2,"realme P4 Lite 5G (Mosaic Green, 64 GB)",Realme,15499,28999.0,46.0,4.3,8068,414,2026-06-30 16:02:20,1,2026-06-30,realme P4 Lite 5G (64 GB)
3,"vivo T5x 5G (Cyber Green, 256 GB)",Vivo,26999,33999.0,20.0,4.4,14728,981,2026-06-30 16:02:20,1,2026-06-30,vivo T5x 5G (256 GB)
4,"vivo T5x 5G (Cyber Green, 128 GB)",Vivo,24999,30999.0,19.0,4.4,14728,981,2026-06-30 16:02:20,1,2026-06-30,vivo T5x 5G (128 GB)
5,"vivo T5 Pro 5G (Glacier Blue, 128 GB)",Vivo,32999,55999.0,41.0,4.5,5729,493,2026-06-30 16:02:20,1,2026-06-30,vivo T5 Pro 5G (128 GB)
...,...,...,...,...,...,...,...,...,...,...,...,...
1318,"POCO M7 Plus 5G (Carbon Black, 128 GB)",POCO,16999,NaN,NaN,4.3,11207,716,2026-07-06 16:54:24,8,2026-07-06,POCO M7 Plus 5G (128 GB)
1323,"Samsung Galaxy S25 5G (Silver Shadow, 128 GB)",Samsung,56999,74999.0,24.0,4.6,4477,483,2026-07-06 16:54:24,8,2026-07-06,Samsung Galaxy S25 5G (128 GB)
1333,"Samsung Galaxy A36 5G (Awesome White, 256 GB)",Samsung,30999,38999.0,20.0,4.3,2543,190,2026-07-06 16:54:24,8,2026-07-06,Samsung Galaxy A36 5G (256 GB)
1334,"realme 15 Pro 5G (Flowing Silver, 128 GB)",Realme,30499,35999.0,15.0,4.3,10443,930,2026-07-06 16:54:24,8,2026-07-06,realme 15 Pro 5G (128 GB)


#### Saving cleaned file

In [41]:
save_path = r"C:\Users\91935\PycharmProjects\flipkart-smartphone-analysis\data\processed\flipkart_smartphones_cleaned.csv"
df_clean.to_csv(save_path, index=False)